Adds nl to IEx.Helpers

bryanhunter opened the pull request:
Erlang's `c` module is a jumbled sock-drawer of commonly used commands. Inside it is a powerful (but obscure) feature `c:nl(modulename)` that will deploy the object code for a locally compiled module to all connected nodes. This makes deploying code to a newly connected cluster simple, and allows easy redeployment of updated/recompiled code. The Erlang implementation of `c:nil` ([doc](http://erlang.org/doc/man/c.html#nl-1)) glues together `code:ensure_loaded` ([doc](http://erlang.org/doc/man/code.html#ensure_loaded-1)) and `rpc:eval_everywhere` ([doc](http://erlang.org/doc/man/rpc.html#eval_everywhere-4)). 

This pull request give the functionality of `nl` a better home: `Node.load_module`

scrogson commented on lib/elixir/lib/node.ex:
```diff
@@ -264,4 +264,28 @@ defmodule Node do
+  def load_module(module) when is_atom(module) do
+    case :code.get_object_code(module) do
+      {^module, bin, beam_path} ->
+        :abcast == :rpc.eval_everywhere(:code, :load_binary, [module, beam_path, bin])
```
Probably meant to just use a single `=` here?

josevalim commented on lib/elixir/lib/node.ex:
```diff
@@ -264,4 +264,28 @@ defmodule Node do
+      {:error, :nofile}
+
+  """
+  @spec load_module(module) :: boolean | {:error, :nofile}
```
What if we make the first argument a list of nodes? Then folks need to call Node.list before but it should be enough.

josevalim replied:
Also, let's make it return `{:ok, boolean} | {:error, :nofile}` or `:ok | {:error, :badrpc} | {:error, :nofile}`. Thoughts?

bryanhunter replied:
Yes, you're right about the node list. Without the explicit node list, it might miss the need of some folks. I considered it, wondered if I was overthinking, and went the chicken route of a straight port of `nl`. :)

Fix coming shortly...

josevalim commented:
Beautiful, I have added some comments. :Heart: I would love to hear @fishcakez's opinion too.

bryanhunter commented:
What do you think of the function name? I considered `deploy_module`. That would be closer to the intent, but it breaks a bit with the language of `Code` and the docs for `c:nl/1`.

josevalim commented:
I like the name. :)

fishcakez commented:
I am not keen on the current (or a similar) implementation living in `Node` because I think that gives the false impression that this is the most suitable method for doing remote deployment of code. I don't think it is because this ignores errors and even if we do a call instead its still possible to end up in limbo where we don't know whether the module was or loaded or not (`:badrpc`). I would think it would rather belong in `IEx.Helpers` - our equivalent to `:c` - for adhoc use, perhaps to load a debugging module or updated code when developing.

I think the only way to provide a reliable mechanism to load remote code would be to pull, rather than push, the code. This already exists with the `inet` loader.

josevalim commented:
So should we move it to `IEx.Helpers.nl/2`?

fishcakez commented:
> So should we move it to IEx.Helpers.nl/2?

I meant that but...

I just remembered watching a talk where it was mentioned they deployed a single application to other nodes like this, I asked them how they handle errors but no reply yet (its been only few minutes). If we can show the result of the load for each node I think we should be good to keep in `Node`, like the return of yield_many.

josevalim commented:
That sounds like a good plan Does this mean we need to map over the nodes and use :rpc.call or something similar?

fishcakez commented:
I think `:rpc.multicall` should be enough?

On 18 February 2016 at 16:04, José Valim <notifications@github.com> wrote:

> That sounds like a good plan Does this mean we need to map over the nodes
> and use :rpc.call or something similar?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/4306#issuecomment-185792392>.
>

fishcakez commented:
Yes, I think using `:rpc.multicall` is the way to go and we can just return the errors and leave it up to the user to handle them.

josevalim commented:
Perfect. So @bryanhunter we just need to move to multicall instead of eval_everywhere and improve the docs on the return values :)

josevalim commented:
Actually, @fishcakez, I think we should still move it to IEx.Helpers, even with those improvements. Let's call it `nl` in there as discussed.

bryanhunter commented:
Great thread! Thanks for all the good thought put into this.

I will add to `IEx.Helpers` a function like:
`def nl(nodes \\ Node.list, module) when is_list(nodes) and is_atom(module) do...`

I really like the idea of `nl` giving feedback on how things went. This will be a nice improvement over Erlang's `c:nl/1`.

One problem. Using `:rpc.multicall` will prevent correlating results (errors and OKs) to the node name where they occurred. For example deploying a good module `Blabber` to three connected nodes via `multicall` returns something like this:
`{[module: Blabber, module: Blabber, module: Blabber], []}`
Not so helpful.

I think I'll need to use `:rpc.call` iteratively to build a return as a list of node names and their results. Does that sound good?

Also, should I submit this as a new pull-request, or simply re-push to this one?

josevalim commented:
That sounds good to me. Also, feel free to repush!

-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

bryanhunter commented:
Moved to IEx.Helpers, renamed to `nl`, adds more tests, and reports responses for each node. Ping me if I've missed something. Thanks for all the great feedback!

whatyouhide commented on lib/elixir/lib/node.ex:
```diff
@@ -264,4 +264,4 @@ defmodule Node do
     :erlang.get_cookie()
   end
-end
+end
```
You deleted the newline at the end of the file here.

whatyouhide commented on lib/iex/lib/iex/helpers.ex:
```diff
@@ -643,4 +644,42 @@ defmodule IEx.Helpers do
+    case :code.get_object_code(module) do
+      {^module, bin, beam_path} ->
+        results = 
+          for node<-nodes do
```
You should probably use spaces around `<-`:

```elixir
for node <- nodes do
```

whatyouhide commented on lib/iex/lib/iex/helpers.ex:
```diff
@@ -643,4 +644,42 @@ defmodule IEx.Helpers do
+      {^module, bin, beam_path} ->
+        results = 
+          for node<-nodes do
+            case :rpc.call(node,:code, :load_binary, [module, beam_path, bin]) do
```
Missing space after `node,`.

whatyouhide commented on lib/iex/lib/iex/helpers.ex:
```diff
@@ -643,4 +644,42 @@ defmodule IEx.Helpers do
+                 {:node1@westhost, :loaded, HelloWorld}]}
+
+      iex> nl(NoSuchModuleExists)
+      {:error, :nofile}
```
I think you used `iex>` just on one of the two examples because only the second one can be easily doctested; I think you shouldn't use `iex>` in the second example either, for consistency. Let's hear what other people think :)

bryanhunter replied:
Aye. Sounds good. Turns out that second test was already represented in helpers_test.

whatyouhide commented on lib/iex/lib/iex/helpers.ex:
```diff
@@ -643,4 +644,42 @@ defmodule IEx.Helpers do
+  Returns `{:error, :nofile}` if the object code (i.e. ".beam" file) for the module
+  could not be found locally.
+
+   ## Examples
```
Misindentation: should be indented on the same level as the rest of the text, it's indented one space more.

whatyouhide commented on lib/iex/test/iex/helpers_test.exs:
```diff
@@ -372,6 +372,13 @@ defmodule IEx.HelpersTest do
+    assert nl(:non_existent_module) == {:error, :nofile}
+    assert nl([node],Enum) == {:ok, [{:nonode@nohost, :loaded, Enum}]}    
+    assert nl([:nosuchnode@badhost],Enum) == {:ok, [{:nosuchnode@badhost, :badrpc, :nodedown}]}
+    capture_log fn -> assert nl([node],:lists) == {:ok, [{:nonode@nohost, :error, :sticky_directory}]} end
```
Missing space after comma, in all three lines above this one.

lexmag replied:
If we move assertion to its own line it improves readability:
```elixir
capture_log fn ->
  assert ...
end
```

whatyouhide commented:
@bryanhunter hey, good work! I left a bunch of style-related comments, sorry for the bombing :)

bryanhunter commented:
@whatyouhide. Wow, apologies for so many! Haha! Thanks for the catches. Updated.

lexmag commented:
@bryanhunter please squash everything into single commit. :bowtie:

bryanhunter commented:
Aye. Squashed :)

lexmag commented on lib/iex/lib/iex/helpers.ex:
```diff
@@ -643,4 +644,42 @@ defmodule IEx.Helpers do
+  ## Examples
+
+      nl(HelloWorld)
+      #=> {:ok, [{:node1@easthost, :loaded, HelloWorld},
```
Very nitpick: maybe we should use actual output from IEx, it has different indentation?
```elixir
{:ok,
 [{:node1@easthost, :loaded, HelloWorld},
  {:node1@westhost, :loaded, HelloWorld}]}
```

bryanhunter replied:
Hi @lexmag. Interesting point.

In the actual output there is no indentation or line-breaks. The return is just a data structure. So the true output would be:  
```
{:ok, [{:node1@easthost, :loaded, HelloWorld}, {:node1@westhost, :loaded, HelloWorld}]}
```
My thinking was the line-breaks and indentation will make it more clear to the reader that for each node there is a response tuple. 

Happy to modify it to the actual output (possibly less effective doc), or to your suggested output (which I like better than my original, but isn't _actual output_ either). 

What do you think is best?

josevalim replied:
Current output is fine. :)

On Saturday, February 27, 2016, Bryan Hunter <notifications@github.com>
wrote:

> In lib/iex/lib/iex/helpers.ex
> <https://github.com/elixir-lang/elixir/pull/4306#discussion_r54336775>:
>
> > +
> > +  @doc """
> > +  Deloys a given module's beam code to a list of connected nodes.
> > +
> > +  This function is useful for development and debugging when you have code that
> > +  has been compiled or updated locally that you want to run on other nodes.
> > +
> > +  The node list defaults to a list of all connected nodes.
> > +
> > +  Returns `{:error, :nofile}` if the object code (i.e. ".beam" file) for the module
> > +  could not be found locally.
> > +
> > +  ## Examples
> > +
> > +      nl(HelloWorld)
> > +      #=> {:ok, [{:node1@easthost, :loaded, HelloWorld},
>
> Hi @lexmag <https://github.com/lexmag>. Interesting point.
>
> In the actual output there is no indentation or line-breaks. The return is
> just a data structure. So the true output would be:
>
> {:ok, [{:node1@easthost, :loaded, HelloWorld}, {:node1@westhost, :loaded, HelloWorld}]}
>
> My thinking was the line-breaks and indentation will make it more clear to
> the reader that for each node there is a response tuple.
>
> Happy to modify it to the actual output (possibly less effective doc), or
> to your suggested output (which I like better than my original, but isn't *actual
> output* either).
>
> What do you think is best?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/4306/files#r54336775>.
>


-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

lexmag replied:
As @josevalim said, current output is totally fine.
I named mine as _actual_ since this is what you see as `IO.inspect/2` output with default options (`width: 80`).

fishcakez commented on lib/iex/lib/iex/helpers.ex:
```diff
@@ -34,6 +34,7 @@ defmodule IEx.Helpers do
     * `l/1`           - loads the given module's beam code
     * `ls/0`          - lists the contents of the current directory
     * `ls/1`          - lists the contents of the specified directory
+    * `nl/2`          - deploys local beam code to a list of connected nodes
```
:scissors: "connected". The node does not need to be connected when the helper is called. If not connected it will try to connect.

josevalim commented:
Merged, thank you!  :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
